Features/vault enterprise

eadderley-tc opened the pull request:
### What does this PR do?
This PR enables Salt to communicate with an Enterprise Vault via X-Vault-Namespace headers.

### What issues does this PR fix or reference?
Fixes: [58585](https://github.com/saltstack/salt/issues/58585)

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

Not entirely sure how to write tests for this, as the headers don't break anything if no namespace is supplied and Vault doesn't complain if namespaces aren't enabled.

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

eadderley-tc committed:
runner, utils and docs

eadderley-tc committed:
runner, utils and docs

eadderley-tc committed:
runner, utils and docs

eadderley-tc committed:
revert that

eadderley-tc committed:
revert that

eadderley-tc committed:
changelog update

eadderley-tc committed:
revert to existing structure to unblock tests

eadderley-tc committed:
revert to existing structure to unblock tests

eadderley-tc committed:
revert to existing structure to unblock tests

mchugh19 commented on salt/utils/vault.py:
```diff
@@ -310,6 +322,17 @@ def make_request(
+        try:
+            namespace = __opts__.get("vault").get("namespace", None)
+        except (TypeError, AttributeError):
+            # Don't worry about setting verify if it doesn't exist
```
Nitpick: Comment doesn't appear to match code (or be needed)

s0undt3ch replied:
Yes, please adjust or remove the comment.

DmitryKuzmenko commented on salt/utils/vault.py:
```diff
@@ -310,6 +322,17 @@ def make_request(
+        except (TypeError, AttributeError):
+            # Don't worry about setting verify if it doesn't exist
+            pass 
+    if namespace is not None:
```
You'll get `"namespace"` is not defined exception here iv `"vault"` not in `__opts__`.

DmitryKuzmenko commented on salt/utils/vault.py:
```diff
@@ -310,6 +322,17 @@ def make_request(
             pass
     url = "{0}/{1}".format(vault_url, resource)
     headers = {"X-Vault-Token": str(token), "Content-Type": "application/json"}
+    # Vault Enterprise requires a namespace
+    if "namespace" in args:
+        namespace = args["namespace"]
+    else:
+        try:
+            namespace = __opts__.get("vault").get("namespace", None)
+        except (TypeError, AttributeError):
+            # Don't worry about setting verify if it doesn't exist
+            pass 
+    if namespace is not None:
+        headers["X-Vault-Namespace"] = namespace
```
```suggestion
    # Vault Enterprise requires a namespace
    try:
        headers["X-Vault-Namespace"] = args.get("namespace") or __opts__["vault"]["namespace"]
    except (TypeError, AttributeError):
        # Don't worry about setting namespace if it doesn't exist
```

DmitryKuzmenko commented:
@eadderley-tc thank you for PR. I've put some comments to my review.
Also about tests: in this case it's needed to add or update the existing tests to cover your changes. I.e. check that the code works as expected in all the cases you've added: there's no namespace specified, there's a namespace specified, exception cases.

eadderley-tc committed:
pulling in Teds fixes

eadderley-tc committed:
pulling in Teds fixes

eadderley-tc committed:
pulling in Teds fixes

eadderley-tc committed:
resolve uninitialized namespace

pmcconnell-tc committed:
Add namespace key to test inputs

pmcconnell-tc committed:
Add namespace key to test inputs

eadderley-tc committed:
remove headers creation without Namespace

eadderley-tc committed:
remove headers creation without Namespace

eadderley-tc committed:
remove headers creation without Namespace

eadderley-tc committed:
Revert "remove headers creation without Namespace"

This reverts commit f72cb25ff5297cacc6121570fdb3a99967f34087.

eadderley-tc committed:
Revert "remove headers creation without Namespace"

This reverts commit f72cb25ff5297cacc6121570fdb3a99967f34087.

eadderley-tc committed:
Revert "remove headers creation without Namespace"

This reverts commit f72cb25ff5297cacc6121570fdb3a99967f34087.

eadderley-tc committed:
headers=ANY to account for namespace logic change

pmcconnell-tc committed:
Add unit tests

pmcconnell-tc committed:
Add unit tests

pmcconnell-tc committed:
Add unit tests

eadderley-tc committed:
lint update

eadderley-tc committed:
lint update

eadderley-tc committed:
utils test fix

eadderley-tc committed:
utils test fix

eadderley-tc commented:
@DmitryKuzmenko We've added some tests and cleaned up the linting. Sorry about the delay, got pulled off for project work and illness. Can you explain what's going wrong in that ci/pre-commit?

sagetherage commented:
@dwoz or @s0undt3ch can you review this PR please? @eadderley-tc Dmitry is unfortunately no longer on the project, but looking to get this reviewed. Pre-commit can be pain, if it is helpful please attend or get with @waynew on IRC, Community Slack or at the Test Clinics we hold on Tue/Thu via Twitch.

s0undt3ch commented on salt/utils/vault.py:
```diff
@@ -151,6 +162,7 @@ def _use_local_config():
                     ]
             return {
                 "url": __opts__["vault"]["url"],
+                "namespace": __opts__["vault"]["namespace"],
```
This is a valid point that needs to be addressed

s0undt3ch commented:
Please remove the changes under `doc/man` those get generated automatically.

And there's also conflicts that need to be solved.

s0undt3ch reviewed (changes requested):
Please remove the changes under doc/man those get generated automatically.

And there's also conflicts that need to be solved.

eadderley-tc commented:
> Please remove the changes under doc/man those get generated automatically.

What do they get automatically generated from?

vveliev-tc committed:
revert doc chnages

vveliev-tc committed:
revert doc changes

vveliev-tc committed:
update docs

vveliev-tc committed:
make lint happy

vveliev-tc committed:
remove py2 dep

vveliev-tc commented:
@s0undt3ch  not sure why some tests are failing, what is currently outstanding for this PR?

vveliev-tc commented:
re-run centos7-py3-pycryptodome-pytest

s0undt3ch commented:
Can you please rebase against the latest master branch changes? Thanks!

vveliev-tc committed:
layout fixes

vveliev-tc commented:
@s0undt3ch, are we good to merge or need to rebase again?

s0undt3ch commented on salt/runners/vault.py:
```diff
@@ -52,6 +52,8 @@ def generate_token(
         config = __opts__.get("vault", {})
         verify = config.get("verify", None)
+        # Vault Enterprise requires a namespace
+        namespace = config.get("namespace", None)
```
```suggestion
        namespace = config.get("namespace")
```
That extra `, None` is not required. The above code does exactly the same

eadderley-tc replied:
Neat. I'm pretty sure I added that for some of the testing, but if that's mechanically true, good to know.

s0undt3ch replied:
yep, it is:
```python
>>> a = {'b': 1}
>>> a.get('b')
1
>>> a.get('c')
>>> a.get('c', None)
>>>
```

s0undt3ch commented on salt/utils/vault.py:
```diff
@@ -103,6 +103,7 @@ def _get_token_and_url_from_master():
         "url": result["url"],
         "token": result["token"],
         "verify": result.get("verify", None),
+        "namespace": result.get("namespace", None),
```
```suggestion
        "namespace": result.get("namespace"),
```

s0undt3ch commented on salt/utils/vault.py:
```diff
@@ -117,6 +118,8 @@ def get_vault_connection():
     def _use_local_config():
         log.debug("Using Vault connection details from local config")
+        # Vault Enterprise requires a namespace
+        namespace = __opts__["vault"].get("namespace", None)
```
```suggestion
        namespace = __opts__["vault"].get("namespace")
```

s0undt3ch commented on salt/utils/vault.py:
```diff
@@ -297,6 +310,7 @@ def make_request(
         connection = get_cache()
     token = connection["token"] if not token else token
     vault_url = connection["url"] if not vault_url else vault_url
+    namespace = connection["namespace"] if not namespace else namespace
```
```suggestion
    namespace = namespace or connection["namespace"]
```

s0undt3ch commented on salt/utils/vault.py:
```diff
@@ -360,10 +376,14 @@ def _selftoken_expired():
     try:
         verify = __opts__["vault"].get("verify", None)
+        # Vault Enterprise requires a namespace
+        namespace = __opts__["vault"].get("namespace", None)
```
```suggestion
        namespace = __opts__["vault"].get("namespace")
```

s0undt3ch commented on salt/utils/vault.py:
```diff
@@ -380,10 +400,14 @@ def _wrapped_token_valid():
     try:
         verify = __opts__["vault"].get("verify", None)
+        # Vault Enterprise requires a namespace
+        namespace = __opts__["vault"].get("namespace", None)
```
```suggestion
        namespace = __opts__["vault"].get("namespace")
```

s0undt3ch commented on tests/integration/files/file/base/Issues_55775.txt.bak:
```diff
@@ -0,0 +1 @@
+foo
```
Why is this file here?

eadderley-tc replied:
What file? :angel:

s0undt3ch commented on tests/unit/runners/test_vault.py:
```diff
@@ -290,6 +282,37 @@ def test_generate_token(self):
+            with patch.dict(vault.__opts__["vault"], supplied_config):
+                result = vault.generate_token("test-minion", "signature")
+                log.debug("generate_token result: %s", result)
+                self.assertTrue(isinstance(result, dict))
```
```suggestion
                self.assertIsInstance(result, dict)
```

s0undt3ch commented on tests/unit/runners/test_vault.py:
```diff
@@ -290,6 +282,37 @@ def test_generate_token(self):
+                result = vault.generate_token("test-minion", "signature")
+                log.debug("generate_token result: %s", result)
+                self.assertTrue(isinstance(result, dict))
+                self.assertFalse("error" in result)
```
```suggestion
                self.assertNotIn("error", result)
```

s0undt3ch commented on tests/unit/runners/test_vault.py:
```diff
@@ -290,6 +282,37 @@ def test_generate_token(self):
+                log.debug("generate_token result: %s", result)
+                self.assertTrue(isinstance(result, dict))
+                self.assertFalse("error" in result)
+                self.assertTrue("token" in result)
```
```suggestion
                self.assertIn("token", result)
```

s0undt3ch reviewed (changes requested):
Sorry. Minor nitpicks(which actually improve readability)

eadderley-tc committed:
Update salt/runners/vault.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

eadderley-tc committed:
Update salt/utils/vault.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

eadderley-tc committed:
Update salt/utils/vault.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

eadderley-tc committed:
Update tests/unit/runners/test_vault.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

eadderley-tc committed:
Update tests/unit/runners/test_vault.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

eadderley-tc committed:
Update tests/unit/runners/test_vault.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

eadderley-tc committed:
Delete Issues_55775.txt.bak

eadderley-tc committed:
Update salt/utils/vault.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

eadderley-tc committed:
Update salt/utils/vault.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

eadderley-tc committed:
Update salt/utils/vault.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

eadderley-tc commented:
@s0undt3ch How's this look?

s0undt3ch commented on salt/runners/vault.py:
```diff
@@ -267,10 +277,15 @@ def _selftoken_expired():
     try:
         verify = __opts__["vault"].get("verify", None)
+        # Vault Enterprise requires a namespace
+        namespace = __opts__["vault"].get("namespace", None)
```
```suggestion
        namespace = __opts__["vault"].get("namespace")
```

s0undt3ch reviewed (changes requested):
Final change and we're good

eadderley-tc committed:
Update salt/runners/vault.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

vveliev-tc commented:
re-run ubuntu1604-py3-tcp-pytest

vveliev-tc commented:
re-run windows2016-py3-pytest

vveliev-tc commented:
re-run windows2019-py3-pytest

vveliev-tc commented:
s0undt3ch - is there anything we should do about failing tests?

eadderley-tc committed:
Merge branch 'master' into features/vault_enterprise

eadderley-tc committed:
Merge branch 'master' into features/vault_enterprise

eadderley-tc commented:
@s0undt3ch As far as we can tell (asked on the Slack too) the remaining failing tests are broken. Can we get another review?

eadderley-tc committed:
Merge branch 'master' into features/vault_enterprise

eadderley-tc commented:
@s0undt3ch Latest master commits merged in

eadderley-tc committed:
Merge branch 'master' into features/vault_enterprise

Ch3LL commented on salt/modules/vault.py:
```diff
@@ -49,6 +50,12 @@
+        Optional Vault Namespace. Used with Vault enterprice
+
+        For detail please see:
+        https://www.vaultproject.io/docs/enterprise/namespaces
```
```suggestion
        https://www.vaultproject.io/docs/enterprise/namespaces
        
        .. versionadded:: 3004
```

eadderley-tc committed:
Update salt/modules/vault.py

Co-authored-by: Megan Wilhite <megan.wilhite@gmail.com>

eadderley-tc committed:
Merge branch 'master' into features/vault_enterprise

Ch3LL reviewed (dismissed):
Looks like there is a just one last lint failure, probably caused by my suggestion. Apologies on that.

eadderley-tc committed:
Update vault.py

Linting

eadderley-tc committed:
Merge branch 'master' into features/vault_enterprise

eadderley-tc committed:
Merge branch 'master' into features/vault_enterprise

eadderley-tc committed:
Merge branch 'master' into features/vault_enterprise

eadderley-tc committed:
Merge branch 'master' into features/vault_enterprise
